File: code/smv-debugger/src/main/java/org/fbme/smvDebugger/model/CounterexampleParser.java
package org.fbme.smvDebugger.model;

import java.util.List;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.Objects;

public class CounterexampleParser {
  private static final String STATE_TRASH_SYMBOLS = "-1[.]";
  private static final String DATA_DELIMITER = ",";

  private final SystemItemParser itemParser = new SystemItemParser();

  public Counterexample parse(final List<String> lines) {
    final String[] emptyHeaderAndSteps = lines.get(0).replaceAll(STATE_TRASH_SYMBOLS, "").split(DATA_DELIMITER);
    final String[] steps = Arrays.copyOfRange(emptyHeaderAndSteps, 1, emptyHeaderAndSteps.length);

    final List<SystemItem> items = new ArrayList<>();
    final List<String[]> values = new ArrayList<>();
    String[] timeValues = new String[steps.length];
    for (int i = 1; i < lines.size(); i++) {
      final String[] lineData = lines.get(i).split(DATA_DELIMITER);
      String fullName = lineData[0];
      final String[] curValues = Arrays.copyOfRange(lineData, 1, lineData.length);

      if (Objects.equals(fullName, Counterexample.T_GLOBAL)) {
        timeValues = curValues;
        continue;
      }

      final SystemItem item = itemParser.parse(fullName);
      if (item != null) {
        items.add(item);
        values.add(curValues);
      }
    }

    return new Counterexample(items.toArray(new SystemItem[0]), steps, values.toArray(new String[0][]), timeValues);
  }
}

File: code/smv-debugger/src/main/kotlin/org/fbme/smvDebugger/model/CounterexampleParser.kt
package org.fbme.smvDebugger.model

import java.util.*

class CounterexampleParser {
    private val itemParser = SystemItemParser()
    fun parse(lines: List<String>): Counterexample {
        val emptyHeaderAndSteps: Array<String> = lines[0].replace(STATE_TRASH_SYMBOLS.toRegex(), "").split(
            DATA_DELIMITER
        ).toTypedArray()
        val steps = emptyHeaderAndSteps.copyOfRange(1, emptyHeaderAndSteps.size)
        val items: MutableList<SystemItem> = ArrayList()
        val values: MutableList<Array<String>> = ArrayList()
        var timeValues = Array(steps.size) { "" }
        for (i in 1 until lines.size) {
            val lineData: Array<String> = lines[i].split(DATA_DELIMITER).toTypedArray()
            val fullName = lineData[0]
            val curValues = lineData.copyOfRange(1, lineData.size)
            if (fullName == Counterexample.T_GLOBAL) {
                timeValues = curValues
                continue
            }
            val item = itemParser.parse(fullName)
            if (item != null) {
                items.add(item)
                values.add(curValues)
            }
        }
        return Counterexample(items.toTypedArray(), steps, values.toTypedArray(), timeValues)
    }

    companion object {
        private const val STATE_TRASH_SYMBOLS = "-1[.]"
        private const val DATA_DELIMITER = ","
    }
}
